File: entity/rag/chunks.py
import uuid
from datetime import datetime

from sqlalchemy import Integer, SmallInteger, String, Text, func, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from entity.base import Base


class Chunk(Base):
    """
    父子块 ORM 模型（映射 rag_chunks）。

    - id 由 PostgreSQL 18 的 uuidv7() 服务端默认生成，level=0 的 id 即 Qdrant point id；
    - document_id / parent_chunk_id 逻辑关联，不加外键/relationship，业务层保证一致性；
    - document_version 冗余存储文档版本，用于重建索引时新旧版本并存/切换；
    - level=0 为叶子子块（入 Qdrant），level>=1 为父块（仅 PG，用于回写上下文）；
    - chunk_index 为同一 (document_id, document_version, level) 内的顺序号。
    """

    __tablename__ = "rag_chunks"
    __table_args__ = {"schema": "rag"}

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("uuidv7()"),
    )

    document_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    parent_chunk_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    document_version: Mapped[int] = mapped_column(Integer, nullable=False)

    level: Mapped[int] = mapped_column(
        SmallInteger, nullable=False, server_default=text("0")
    )
    chunk_index: Mapped[int] = mapped_column(Integer, nullable=False)

    content: Mapped[str] = mapped_column(Text, nullable=False)
    token_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    char_start: Mapped[int | None] = mapped_column(Integer, nullable=True)
    char_end: Mapped[int | None] = mapped_column(Integer, nullable=True)

    chapter_title: Mapped[str | None] = mapped_column(Text, nullable=True)
    page_start: Mapped[int | None] = mapped_column(Integer, nullable=True)
    page_end: Mapped[int | None] = mapped_column(Integer, nullable=True)

    content_hash: Mapped[str | None] = mapped_column(String(64), nullable=True)
    chunk_metadata: Mapped[dict] = mapped_column(
        "metadata", JSONB, nullable=False, server_default=text("'{}'::jsonb")
    )

    created_at: Mapped[datetime] = mapped_column(
        nullable=False, server_default=func.now()
    )

    def to_dict(self) -> dict:
        """转为可 JSON 序列化的普通字典，供统一响应回写。"""
        return {
            "id": str(self.id),
            "document_id": str(self.document_id),
            "parent_chunk_id": (
                str(self.parent_chunk_id) if self.parent_chunk_id else None
            ),
            "document_version": self.document_version,
            "level": self.level,
            "chunk_index": self.chunk_index,
            "content": self.content,
            "token_count": self.token_count,
            "char_start": self.char_start,
            "char_end": self.char_end,
            "chapter_title": self.chapter_title,
            "page_start": self.page_start,
            "page_end": self.page_end,
            "content_hash": self.content_hash,
            "metadata": self.chunk_metadata,
            "created_at": self.created_at,
        }

File: multi-agent-ui/src/api/rag/chunks.ts
import request, { type ApiResult } from "@/utils/request";
import { CHUNK_URL, type PageResult } from "./index";

/** 分块实体（对应后端 rag_chunks.to_dict） */
export interface Chunk {
  id: string;
  /** 归属组织/部门（逻辑指向 sys_dept.id，空字符串表示未归属），由服务端注入，不可变 */
  dept_id: string;
  document_id: string;
  parent_chunk_id: string | null;
  document_version: number;
  /** 0 叶块（入 Qdrant） / 1 父块（仅 PG，用于回写上下文） */
  level: number;
  chunk_index: number;
  content: string;
  token_count: number | null;
  char_start: number | null;
  char_end: number | null;
  chapter_title: string | null;
  page_start: number | null;
  page_end: number | null;
  content_hash: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

/** 分页列出分块参数 */
export interface ChunkListParams {
  /** 按文档过滤 */
  document_id: string;
  /** 按层级过滤（0 叶块 / 1 父块） */
  level?: number;
  /** 按文档版本过滤 */
  document_version?: number;
  /** 页码，从 1 开始（默认 1） */
  page?: number;
  /** 每页数量（1-200，默认 50） */
  size?: number;
}

/** 分块管理 API：统一通过 chunkApi.xx() 调用 */
export const chunkApi = {
  /** 按文档分页列出分块，可选 level/document_version 过滤，按 chunk_index 升序 */
  list(params: ChunkListParams) {
    return request.get<PageResult<Chunk>, ApiResult<PageResult<Chunk>>>(CHUNK_URL.root, {
      params,
    });
  },

  /** 按 id 获取分块 */
  get(chunkId: string) {
    return request.get<Chunk, ApiResult<Chunk>>(CHUNK_URL.byId(chunkId));
  },
};
